rust-runtime/aws-smithy-checksums/src/body.rs

```diff
@@ -0,0 +1,444 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::http::{new_from_algorithm, HttpChecksum};
+
+use aws_smithy_http::body::SdkBody;
+use aws_smithy_http::header::append_merge_header_maps;
+use aws_smithy_types::base64;
+
+use bytes::Bytes;
+use http::header::HeaderName;
+use http::{HeaderMap, HeaderValue};
+use http_body::{Body, SizeHint};
+use pin_project_lite::pin_project;
+
+use std::fmt::Display;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+// pin_project_lite can't handle triple-slash doc comments
+pin_project! {
+    /// A `ChecksumBody` will read and calculate a request body as it's being sent. Once the body has"
+    /// been completely read, it'll append a trailer with the calculated checksum."
+    pub struct ChecksumBody<InnerBody> {
+        #[pin]
+        inner: InnerBody,
+        #[pin]
+        checksum: Box<dyn HttpChecksum>,
```

Comment:
`HttpChecksum` doesn't have any functions on it that have a `Pin` self type, so this shouldn't need pin projection, right?

Reply:
I'm pretty foggy on this actually. As far as I understand it, because I have to mutate both `ChecksumBody.inner` and `ChecksumBody.checksum` in the poll, I have to pin them. Perhaps @rcoh could explain it in a more correct and helpful way.

Reply:
Box is Unpin so I don't think you need to pin this